pkg/tinydtls/contrib/sock_dtls.c

```diff
@@ -116,7 +127,26 @@ static int _event(struct dtls_context_t *ctx, session_t *session,
             break;
     }
 #endif  /* ENABLE_DEBUG */
-    mbox_put(&sock->mbox, &msg);
+    if (code != DTLS_EVENT_CONNECT) {
+        mbox_put(&sock->mbox, &msg);
+    }
```

Comment:
```suggestion
    /* only put msg in mbox if successfully connected, skipping alert (`level == 0`) and other events. */
    if (!level && code == DTLS_EVENT_CONNECTED) {
        mbox_put(&sock->mbox, &msg);
    }
```

Reply:
That fixes at least the problem of initializing another handshake. However, when I remove the global session variable (see a4a5483), I still run into the `-ENOTCONN` problem on send on server side :-/.

Reply:
Patch for the fix here: https://gist.github.com/pokgak/e4a13c84a3ab8d693a3a64b319921771

The problem was when the decrypted data is ready, we only save the data, and not the session that it came from. This patch adds the session and returns it from `sock_dtls_recv` when `_copy_buffer` is called.

This is a minimal fix but I suggest we group together the `sock->buf` and the saved session together in a struct like so:

```C
# sock_dtls_types.h
struct sock_dtls {
...
    /**
     * @brief Buffer used to pass decrypted data and its session information.
     */
    struct {
        uint8_t *data;    // renamed from sock->buf
        size_t datalen;
        session_t *session;
    } buffer;
}
```

Reply:
You want to make a PR for that or shall I? This bug also exists in master consequently.

Reply:
I'll make the PR :)

Reply:
Maybe put https://github.com/RIOT-OS/RIOT/pull/12907/commits/0b729ff8012bbc17cca2dccb402f841af2480830 in there as well. IMHO, this behavior is also a problem in master.